fvm/evm/offchain/blocks/blocks.go

```diff
@@ -0,0 +1,122 @@
+package blocks
+
+import (
+	"fmt"
+
+	gethCommon "github.com/onflow/go-ethereum/common"
+
+	"github.com/onflow/flow-go/fvm/evm/handler"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const BlockStoreLatestBlockMetaKey = "LatestBlockMeta"
+
+// Blocks facilitates access to the recent block hash values
+// and also the latest executed block meta data
+type Blocks struct {
+	chainID     flow.ChainID
+	storage     types.BackendStorage
+	rootAddress flow.Address
+	bhl         *handler.BlockHashList
+}
+
+// NewBlocks constructs a new blocks type
+func NewBlocks(
+	chainID flow.ChainID,
+	rootAddress flow.Address,
+	storage types.BackendStorage,
+) (*Blocks, error) {
+	var err error
+	blocks := &Blocks{
+		chainID:     chainID,
+		storage:     storage,
+		rootAddress: rootAddress,
+	}
+	blocks.bhl, err = handler.NewBlockHashList(
+		storage,
+		rootAddress,
+		handler.BlockHashListCapacity,
+	)
+	if err != nil {
+		return nil, err
+	}
+	// if empty insert genesis block hash
+	if blocks.bhl.IsEmpty() {
+		genesis := types.GenesisBlock(chainID)
+		err = blocks.PushBlockMeta(
+			NewMeta(
+				genesis.Height,
+				genesis.Timestamp,
+				genesis.PrevRandao,
+			))
+		if err != nil {
+			return nil, err
+		}
+		// push block hash
+		err = blocks.PushBlockHash(
+			genesis.Height,
+			types.GenesisBlockHash(chainID))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return blocks, nil
+}
+
+// PushBlock pushes a new block into the storage
+func (b *Blocks) PushBlockMeta(
+	meta *Meta,
+) error {
+	// check height order
+	if meta.Height > 0 {
+		bm, err := b.LatestBlock()
+		if err != nil {
+			return err
+		}
+		if meta.Height != bm.Height+1 {
+			return fmt.Errorf("out of order block meta push! got: %d, expected %d ", meta.Height, bm.Height+1)
+		}
+	}
+	return b.storeBlockMetaData(meta)
+}
+
+// PushBlockHash pushes a new block block hash into the storage
+func (b *Blocks) PushBlockHash(
+	height uint64,
+	hash gethCommon.Hash,
+) error {
+	return b.bhl.Push(height, hash)
```

Comment:
Do we need to make sure the meta data for the previous block has been received? Because otherwise we won't be able to push meta for the previous block.

For instance, we expect the following order: 
```
PushBlockMeta(3)
PushBlockHash(3, A)
PushBlockMeta(4)
PushBlockHash(4, B)
```

But what prevents them to be called in this order?
```
PushBlockMeta(3)
PushBlockHash(3, A)
PushBlockHash(4, B)
PushBlockMeta(4) <- this will fail because the latest block is block 4 now
```

Reply:
gonna address this in the following PR.